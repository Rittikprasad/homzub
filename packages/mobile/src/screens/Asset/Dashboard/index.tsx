import React from 'react';
import { StyleProp, ViewStyle } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { SvgProps } from 'react-native-svg';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { ErrorUtils } from '@homzhub/common/src/utils/ErrorUtils';
import { FunctionUtils } from '@homzhub/common/src/utils/FunctionUtils';
import { AnalyticsService } from '@homzhub/common/src/services/Analytics/AnalyticsService';
import { ObjectMapper } from '@homzhub/common/src/utils/ObjectMapper';
import { EventType } from '@homzhub/common/src/services/Analytics/EventType';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { DashboardRepository } from '@homzhub/common/src/domain/repositories/DashboardRepository';
import { AssetActions } from '@homzhub/common/src/modules/asset/actions';
import { PortfolioActions } from '@homzhub/common/src/modules/portfolio/actions';
import { RecordAssetActions } from '@homzhub/common/src/modules/recordAsset/actions';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import { SearchActions } from '@homzhub/common/src/modules/search/actions';
import { CommonSelectors } from '@homzhub/common/src/modules/common/selectors';
import { SearchSelector } from '@homzhub/common/src/modules/search/selectors';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import AddProperty from '@homzhub/common/src/assets/images/addProperty.svg';
import Alert from '@homzhub/common/src/assets/images/alert.svg';
import CustomSearch from '@homzhub/common/src/assets/images/customSearch.svg';
import Finance from '@homzhub/common/src/assets/images/finance.svg';
import Help from '@homzhub/common/src/assets/images/headphone.svg';
import ServiceRequest from '@homzhub/common/src/assets/images/serviceRequest.svg';
import {
  AssetAdvertisementBanner,
  AssetMetricsList,
  AssetSummary,
  FullScreenAssetDetailsCarousel,
} from '@homzhub/mobile/src/components';
import AssetMarketTrends from '@homzhub/mobile/src/components/molecules/AssetMarketTrends';
import IconSheet, { ISheetData } from '@homzhub/mobile/src/components/molecules/IconSheet';
import UserSubscriptionPlan from '@homzhub/common/src/components/molecules/UserSubscriptionPlan';
import WelcomeCard from '@homzhub/common/src/components/molecules/WelcomeCard';
import FinanceOverview from '@homzhub/mobile/src/components/organisms/FinanceOverview';
import LeasePropertyList from '@homzhub/mobile/src/components/organisms/LeasePropertyList';
import PendingPropertyListCard from '@homzhub/mobile/src/components/organisms/PendingPropertyListCard';
import VisitListCard from '@homzhub/mobile/src/components/organisms/VisitListCard';
import { UserScreen } from '@homzhub/mobile/src/components/HOC/UserScreen';
import { Asset, PropertyStatus } from '@homzhub/common/src/domain/models/Asset';
import { Filters } from '@homzhub/common/src/domain/models/AssetFilter';
import { IActions, ISelectedAssetPlan } from '@homzhub/common/src/domain/models/AssetPlan';
import { IMetricsData } from '@homzhub/common/src/domain/models/AssetMetrics';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { DashboardNavigatorParamList } from '@homzhub/mobile/src/navigation/DashboardStack';
import { Attachment } from '@homzhub/common/src/domain/models/Attachment';
import { Currency } from '@homzhub/common/src/domain/models/Currency';
import { IFilter } from '@homzhub/common/src/domain/models/Search';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { ISetAssetPayload } from '@homzhub/common/src/modules/portfolio/interfaces';
import { IAssetVisitPayload, VisitStatus } from '@homzhub/common/src/domain/repositories/interfaces';
import { IApiClientError } from '@homzhub/common/src/network/ApiClientError';

interface IDispatchProps {
  setCurrentFilter: (payload: Filters) => void;
  setAssetId: (payload: number) => void;
  setSelectedPlan: (payload: ISelectedAssetPlan) => void;
  setAddPropertyFlow: (payload: boolean) => void;
  setCurrentAsset: (payload: ISetAssetPayload) => void;
  getAssetVisit: (payload: IAssetVisitPayload) => void;
  setInitialState: () => void;
  getTenanciesDetails: () => void;
  getFilterDetails: (payload: IFilter) => void;
  setFilter: (payload: IFilter) => void;
  setEditPropertyFlow: (payload: boolean) => void;
  toggleEditPropertyFlowBottomSheet: (payload: boolean) => void;
}

interface IReduxStateProps {
  assets: Asset[];
  filters: IFilter;
  defaultCurrency: Currency;
}

type libraryProps = NavigationScreenProps<DashboardNavigatorParamList, ScreensKeys.DashboardLandingScreen>;
type Props = WithTranslation & libraryProps & IDispatchProps & IReduxStateProps;

interface IDashboardState {
  metrics: IMetricsData;
  pendingProperties: Asset[];
  isLoading: boolean;
  showBottomSheet: boolean;
}

enum MetricType {
  OFFER = 'OFFER',
  SITE_VISITS = 'SITE_VISIT',
  SHORTLISTED = 'SHORTLISTED',
}

const ShowInMvpRelease = false;

export class Dashboard extends React.PureComponent<Props, IDashboardState> {
  public focusListener: any;
  public blurListener: any;

  public state = {
    metrics: {} as IMetricsData,
    pendingProperties: [],
    isLoading: false,
    showBottomSheet: false,
  };

  public componentDidMount = (): void => {
    const { navigation } = this.props;
    this.focusListener = navigation.addListener('focus', () => {
      this.getScreenData().then();
    });
    this.blurListener = navigation.addListener('blur', () => {
      this.handleScreenBlur();
    });
  };

  public componentWillUnmount(): void {
    this.focusListener();
    this.blurListener();
  }

  public render = (): React.ReactElement => {
    const { t, route } = this.props;
    const { isLoading, pendingProperties, metrics } = this.state;

    return (
      <>
        <UserScreen loading={isLoading} isGradient title={t('dashboard')} onPlusIconClicked={this.openBottomSheet}>
          {this.renderAssetMetricsAndUpdates()}
          {pendingProperties.length > 0 && (
            <PendingPropertyListCard
              data={pendingProperties}
              onPressComplete={this.onCompleteDetails}
              onSelectAction={this.handleActionSelection}
              onViewProperty={this.onViewProperty}
            />
          )}
          {!metrics?.metricValues?.assets?.count && <WelcomeCard handlePress={this.handleAddProperty} />}
          {metrics?.isTenant && <LeasePropertyList />}
          {metrics?.isTenant && <VisitListCard />}
          {!metrics?.isTenant && <FinanceOverview />}
          <AssetMarketTrends isDashboard onViewAll={this.onViewAll} onTrendPress={this.onTrendPress} />
          <AssetAdvertisementBanner />
          {ShowInMvpRelease && <UserSubscriptionPlan onApiFailure={this.onAssetSubscriptionApiFailure} />}
          {this.renderBottomSheet()}
        </UserScreen>
        {Boolean(route?.params?.imageLink && route?.params?.imageLink?.length > 0) && this.renderFullImageView()}
      </>
    );
  };

  public renderAssetMetricsAndUpdates = (): React.ReactElement => {
    const { t } = this.props;
    const { metrics } = this.state;
    return (
      <>
        <AssetMetricsList
          title={`${metrics?.metricValues?.assets?.count ?? 0}`}
          data={metrics?.metricValues?.miscellaneous ?? []}
          subscription={metrics?.metricValues?.userServicePlan?.label}
          onMetricsClicked={this.handleMetricsNavigation}
          subTitleText={metrics?.isTenant ? t('assetPortfolio:tenancies') : t('common:properties')}
        />
        <AssetSummary
          notification={metrics?.updates?.notifications?.count ?? 0}
          serviceTickets={metrics?.updates?.tickets?.open?.count ?? 0}
          dues={metrics?.updates?.dues?.count ?? 0}
          messages={metrics?.updates?.messages?.unread?.count ?? 0}
          containerStyle={styles.assetCards()}
          onPressServiceTickets={this.handleServiceTickets}
          onPressNotification={this.handleNotification}
          onPressMessages={this.handleMessages}
        />
      </>
    );
  };

  public renderBottomSheet = (): React.ReactElement => {
    const { showBottomSheet } = this.state;
    const data = this.formatBottomSheetData();

    return <IconSheet data={data} onCloseSheet={this.closeBottomSheet} isVisible={showBottomSheet} numOfColumns={3} />;
  };

  public renderFullImageView = (): React.ReactElement | null => {
    const { route } = this.props;

    if (route?.params?.imageLink) {
      const imageLink = ObjectMapper.deserialize(Attachment, { link: route?.params?.imageLink });

      const onPressCross = (): void => {
        this.handleScreenBlur();
      };

      return (
        <>
          <FullScreenAssetDetailsCarousel
            data={[imageLink]}
            activeSlide={0}
            onFullScreenToggle={onPressCross}
            hasOnlyImages
          />
        </>
      );
    }

    return null;
  };

  // HANDLERS
  private onAssetSubscriptionApiFailure = (err: IApiClientError): void => {
    AlertHelper.error({ message: ErrorUtils.getErrorMessage(err) });
  };

  private onCompleteDetails = (assetId: number): void => {
    const { navigation, setAssetId, setEditPropertyFlow, toggleEditPropertyFlowBottomSheet } = this.props;
    setAssetId(assetId);
    setEditPropertyFlow(true);
    toggleEditPropertyFlowBottomSheet(true);
    // @ts-ignore
    navigation.navigate(ScreensKeys.PropertyPostStack, {
      screen: ScreensKeys.AddProperty,
      params: { previousScreen: ScreensKeys.Dashboard },
    });
  };

  private onViewAll = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.MarketTrends, { isFromDashboard: true });
  };

  private onTrendPress = (url: string, trendId: number): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.WebViewScreen, { url, trendId });
  };

  private onViewProperty = (data: ISetAssetPayload): void => {
    const { setCurrentAsset, navigation } = this.props;
    setCurrentAsset(data);
    navigation.navigate(ScreensKeys.PropertyDetailScreen);
  };

  private handleScreenBlur = (): void => {
    const { navigation } = this.props;
    navigation.setParams({ imageLink: undefined });
  };

  private handleMessages = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.Messages, { isFromDashboard: true });
  };

  private handleServiceTickets = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.ServiceTicketScreen, { isFromDashboard: true });
  };

  private handleAddProperty = (): void => {
    const { navigation } = this.props;
    // @ts-ignore
    navigation.navigate(ScreensKeys.PropertyPostStack, { screen: ScreensKeys.AssetLocationSearch });
    AnalyticsService.track(EventType.AddPropertyInitiation);
  };

  private handleNotification = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.AssetNotifications, { isFromDashboard: true });
  };

  private handleMetricsNavigation = (name: string): void => {
    const { navigation, setCurrentFilter, t } = this.props;

    const handleNavigation = (): void => {
      const callNavigation = (screen: ScreensKeys): void =>
        // @ts-ignore
        navigation.navigate(screen, { screenTitle: t('assetDashboard:dashboard') });

      switch (name) {
        case MetricType.OFFER:
          callNavigation(ScreensKeys.PropertyOfferList);
          break;
        case MetricType.SHORTLISTED:
          callNavigation(ScreensKeys.SavedPropertiesScreen);
          break;
        case MetricType.SITE_VISITS:
          callNavigation(ScreensKeys.PropertyVisits);
          break;
        default:
          FunctionUtils.noop();
      }
    };

    if ([MetricType.OFFER, MetricType.SHORTLISTED, MetricType.SITE_VISITS].includes(name as MetricType)) {
      handleNavigation();
      return;
    }
    setCurrentFilter(name as Filters);
    // @ts-ignore
    navigation.navigate(ScreensKeys.Portfolio, {
      screen: ScreensKeys.PortfolioLandingScreen,
      params: {
        isFromNavigation: true,
      },
    });
  };

  private handleActionSelection = (item: IActions, assetId: number): void => {
    const { navigation, setSelectedPlan, setAssetId } = this.props;
    setSelectedPlan({ id: item.id, selectedPlan: item.type });
    setAssetId(assetId);
    // @ts-ignore
    navigation.navigate(ScreensKeys.PropertyPostStack, {
      screen: ScreensKeys.AssetListing,
      params: { previousScreen: ScreensKeys.Dashboard },
    });
  };

  private openBottomSheet = (): void => this.setState({ showBottomSheet: true });

  private closeBottomSheet = (): void => {
    const { setInitialState } = this.props;
    setInitialState();
    this.setState({ showBottomSheet: false });
  };

  private formatBottomSheetData = (): ISheetData[] => {
    const { t, navigation, assets, getFilterDetails, filters, setFilter, defaultCurrency } = this.props;
    const iconSize = 38;

    const handleAddFinancialRecord = (): void => {
      if (assets.length <= 0) {
        AlertHelper.error({ message: t('assetFinancial:addProperty') });
        return;
      }
      navigation.navigate(ScreensKeys.AddRecordScreen, { isFromDashboard: true });
    };

    const handleAddServiceTicket = (): void => {
      if (assets.length <= 0) {
        AlertHelper.error({ message: t('assetDashboard:addPropertyForTicket') });
        return;
      }
      navigation.navigate(ScreensKeys.AddServiceTicket, { isFromDashboard: true });
    };

    const handleAddSupportTicket = (isBugReport?: boolean): void => {
      navigation.navigate(ScreensKeys.SupportScreen, {
        isFromDashboard: true,
        ...(isBugReport && { categoryId: 41 }), // For Bug Report hard-coding the bug category id
      });
    };

    const handleCustomSearch = (): void => {
      getFilterDetails({ asset_group: filters.asset_group });
      setFilter({ currency_code: defaultCurrency.currencyCode });
      navigation.navigate(ScreensKeys.SearchRequirement, { isFromAuth: false });
    };

    const ImageHOC = (Image: React.FC<SvgProps>): React.ReactElement => <Image width={iconSize} height={iconSize} />;

    return [
      {
        icon: ImageHOC(AddProperty),
        label: t('property:addProperty'),
        onPress: this.handleAddProperty,
      },
      {
        icon: ImageHOC(Finance),
        label: t('assetFinancial:addFinancials'),
        onPress: handleAddFinancialRecord,
      },
      {
        icon: ImageHOC(ServiceRequest),
        label: t('assetDashboard:serviceTicket'),
        onPress: handleAddServiceTicket,
      },
      {
        icon: ImageHOC(CustomSearch),
        label: t('propertySearch:shareRequirement'),
        onPress: handleCustomSearch,
      },
      {
        icon: ImageHOC(Help),
        label: t('common:askHelp'),
        onPress: handleAddSupportTicket,
      },
      {
        icon: ImageHOC(Alert),
        label: t('common:bugReport'),
        onPress: (): void => handleAddSupportTicket(true),
      },
    ];
  };

  // HANDLERS end

  // APIs
  private getScreenData = async (): Promise<void> => {
    const { getTenanciesDetails } = this.props;
    await this.getAssetMetrics();
    await this.getPendingProperties();
    this.getVisitData();
    getTenanciesDetails();
  };

  private getAssetMetrics = async (): Promise<void> => {
    this.setState({ isLoading: true });
    try {
      const { metricsData } = await DashboardRepository.getAssetMetrics('v4');
      this.setState({ metrics: metricsData, isLoading: false });
    } catch (e) {
      this.setState({ isLoading: false });
      const error = ErrorUtils.getErrorMessage(e.details);
      AlertHelper.error({ message: error });
    }
  };

  private getPendingProperties = async (): Promise<void> => {
    try {
      const response: Asset[] = await AssetRepository.getPropertiesByStatus(PropertyStatus.PENDING);
      this.setState({ pendingProperties: response });
    } catch (e) {
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details) });
    }
  };

  private getVisitData = (): void => {
    const { getAssetVisit } = this.props;
    getAssetVisit({
      start_date__gte: new Date().toISOString(),
      status__in: `${VisitStatus.APPROVED},${VisitStatus.PENDING}`,
    });
  };
  // APIs end
}

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { setCurrentFilter, setCurrentAsset, setInitialState, getTenanciesDetails } = PortfolioActions;
  const { setAddPropertyFlow } = UserActions;
  const { setAssetId, setSelectedPlan, setEditPropertyFlow, toggleEditPropertyFlowBottomSheet } = RecordAssetActions;
  const { getAssetVisit } = AssetActions;
  const { getFilterDetails, setFilter } = SearchActions;
  return bindActionCreators(
    {
      setCurrentFilter,
      setAssetId,
      setSelectedPlan,
      setAddPropertyFlow,
      setCurrentAsset,
      setInitialState,
      getAssetVisit,
      getTenanciesDetails,
      getFilterDetails,
      setFilter,
      setEditPropertyFlow,
      toggleEditPropertyFlowBottomSheet,
    },
    dispatch
  );
};

const mapStateToProps = (state: IState): IReduxStateProps => {
  const { getUserAssets } = UserSelector;
  const { getFilters } = SearchSelector;
  const { getDefaultCurrency } = CommonSelectors;
  return {
    assets: getUserAssets(state),
    filters: getFilters(state),
    defaultCurrency: getDefaultCurrency(state),
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation(LocaleConstants.namespacesKey.assetDashboard)(Dashboard));

const styles = {
  assetCards: (): StyleProp<ViewStyle> => ({
    marginVertical: 12,
  }),
  evenItem: (): StyleProp<ViewStyle> => ({
    marginEnd: 17,
  }),
};
