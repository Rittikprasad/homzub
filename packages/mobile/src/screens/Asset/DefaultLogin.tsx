import React from 'react';
import { View, StyleSheet } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { theme } from '@homzhub/common/src/styles/theme';
import Logo from '@homzhub/common/src/assets/images/homzhubLogo.svg';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { BottomTabNavigatorParamList } from '@homzhub/mobile/src/navigation/BottomTabs';
import { Button } from '@homzhub/common/src/components/atoms/Button';
import { Text } from '@homzhub/common/src/components/atoms/Text';

type libraryProps = WithTranslation & NavigationScreenProps<BottomTabNavigatorParamList, ScreensKeys.DefaultLogin>;
type IProps = libraryProps;

export class DefaultLogin extends React.PureComponent<IProps> {
  public render(): React.ReactElement {
    const { t } = this.props;
    return (
      <View style={styles.container}>
        <Logo height={100} width={100} />
        <Text type="small" textType="regular" style={styles.loginText}>
          {t('pleaseSignup')}
        </Text>
        <Button type="primary" title={t('signUp')} containerStyle={styles.button} onPress={this.navigateToSignup} />
      </View>
    );
  }

  public navigateToSignup = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.AuthStack, {
      screen: ScreensKeys.SignUp,
      params: { onCallback: this.navigateToDashboard },
    });
  };

  private navigateToDashboard = (): void => {
    const { navigation } = this.props;
    // @ts-ignore
    navigation.navigate(ScreensKeys.BottomTabs);
  };
}

export default withTranslation()(DefaultLogin);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: theme.colors.white,
  },
  loginText: {
    marginVertical: 10,
  },
  button: {
    flex: 0,
    marginVertical: 10,
  },
});
