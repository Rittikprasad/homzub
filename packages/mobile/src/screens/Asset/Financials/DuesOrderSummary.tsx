import React from 'react';
import { StyleSheet } from 'react-native';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';
import { useNavigation } from '@react-navigation/native';
import { theme } from '@homzhub/common/src/styles/theme';
import { FinancialSelectors } from '@homzhub/common/src/modules/financials/selectors';
import { Screen } from '@homzhub/mobile/src/components/HOC/Screen';
import ServiceOrderSummary from '@homzhub/common/src/components/organisms/ServiceOrderSummary';
import { ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';

const DuesOrderSummary = (): React.ReactElement | null => {
  const { goBack, navigate } = useNavigation();
  const { t } = useTranslation();
  const currentDueId = useSelector(FinancialSelectors.getCurrentDueId);
  const { dueOrderSummary } = useSelector(FinancialSelectors.getFinancialLoaders);

  const onSuccess = (): void => {
    navigate(ScreensKeys.FinancialsLandingScreen);
  };

  const onBackIconPress = (): void => {
    goBack();
  };

  return (
    <Screen
      headerProps={{
        title: t('property:orderSummary'),
        onIconPress: onBackIconPress,
      }}
      contentContainerStyle={styles.screenContentContainer}
      scrollEnabled
      isLoading={dueOrderSummary}
    >
      <ServiceOrderSummary isLabelRequired invoiceId={currentDueId} onSuccess={onSuccess} />
    </Screen>
  );
};

export default DuesOrderSummary;

const styles = StyleSheet.create({
  screenContentContainer: {
    paddingVertical: 20,
    paddingHorizontal: 0,
    backgroundColor: theme.colors.white,
  },
});
