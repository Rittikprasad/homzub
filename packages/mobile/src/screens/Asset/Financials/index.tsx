import React from 'react';
import { PickerItemProps } from 'react-native';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import { withTranslation, WithTranslation } from 'react-i18next';
import { uniqBy } from 'lodash';
import { SvgProps } from 'react-native-svg';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { DateUtils } from '@homzhub/common/src/utils/DateUtils';
import { AssetActions } from '@homzhub/common/src/modules/asset/actions';
import { FinancialActions } from '@homzhub/common/src/modules/financials/actions';
import { PropertyPaymentActions } from '@homzhub/common/src/modules/propertyPayment/actions';
import { FinancialSelectors } from '@homzhub/common/src/modules/financials/selectors';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { theme } from '@homzhub/common/src/styles/theme';
import Accounting from '@homzhub/common/src/assets/images/accounting.svg';
import Bank from '@homzhub/common/src/assets/images/bank.svg';
import Payment from '@homzhub/common/src/assets/images/payment.svg';
import { AssetMetricsList, IMetricsData } from '@homzhub/mobile/src/components';
import { UserScreen } from '@homzhub/mobile/src/components/HOC/UserScreen';
import IconSheet, { ISheetData } from '@homzhub/mobile/src/components/molecules/IconSheet';
import { PropertyByCountryDropdown } from '@homzhub/common/src/components/molecules/PropertyByCountryDropdown';
import FinanceOverview from '@homzhub/mobile/src/components/organisms/FinanceOverview';
import DuesContainer from '@homzhub/mobile/src/components/organisms/DuesContainer';
import RemindersContainer from '@homzhub/mobile/src/components/organisms/RemindersContainer';
import TransactionsContainer from '@homzhub/mobile/src/components/organisms/TransactionsContainer';
import { FinancialsNavigatorParamList } from '@homzhub/mobile/src/navigation/FinancialStack';
import { Asset } from '@homzhub/common/src/domain/models/Asset';
import { Country } from '@homzhub/common/src/domain/models/Country';
import { GeneralLedgers } from '@homzhub/common/src/domain/models/GeneralLedgers';
import { User } from '@homzhub/common/src/domain/models/User';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { IFinancialState, ILedgerMetrics } from '@homzhub/common/src/modules/financials/interfaces';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';

interface IOwnState {
  isLoading: boolean;
  showAddSheet: boolean;
}

interface IStateProps {
  assets: Asset[];
  financialLoaders: IFinancialState['loaders'];
  ledgerData: GeneralLedgers[];
  selectedProperty: number;
  selectedCountry: number;
  ledgerMetrics: ILedgerMetrics;
}

interface IDispatchProps {
  getLedgers: () => void;
  setCurrentCountry: (country: number) => void;
  setCurrentProperty: (property: number) => void;
  setTimeRange: (range: number) => void;
  getLedgerMetrics: () => void;
  resetLedgerFilters: () => void;
  setCurrentDueId: (dueId: number) => void;
  setCurrentReminderId: (id: number) => void;
  getAssetUsersSuccess: (payload: User[]) => void;
  clearReminderFormData: () => void;
  resetPaymentState: () => void;
}

type libraryProps = NavigationScreenProps<FinancialsNavigatorParamList, ScreensKeys.FinancialsLandingScreen>;
type Props = WithTranslation & libraryProps & IStateProps & IDispatchProps;

export class Financials extends React.PureComponent<Props, IOwnState> {
  private onFocusSubscription: any;

  public state = {
    isLoading: false,
    showAddSheet: false,
  };

  public componentDidMount(): void {
    const {
      navigation,
      getLedgerMetrics,
      resetLedgerFilters,
      setCurrentDueId,
      setCurrentReminderId,
      getAssetUsersSuccess,
      clearReminderFormData,
      resetPaymentState,
    } = this.props;

    this.onFocusSubscription = navigation.addListener('focus', (): void => {
      resetLedgerFilters();
      getLedgerMetrics();
      setCurrentDueId(-1);
      setCurrentReminderId(-1);
      getAssetUsersSuccess([]);
      clearReminderFormData();
      resetPaymentState();
    });
  }

  public render = (): React.ReactElement => {
    const {
      t,
      route: { params },
      financialLoaders: { dues, payment, reminder },
      selectedCountry,
      selectedProperty,
    } = this.props;
    const { isLoading } = this.state;

    const loading = isLoading || dues || payment || reminder;

    const toggleAddSheet = (): void => this.toggleAddSheet(true);

    return (
      <UserScreen loading={loading} isGradient title={t('financial')} onPlusIconClicked={toggleAddSheet}>
        <AssetMetricsList
          title={t('assetFinancial:summary')}
          numOfElements={2}
          showBackIcon={params?.isFromNavigation ?? false}
          isSubTextRequired={false}
          data={this.getHeaderData()}
        />
        <PropertyByCountryDropdown
          selectedProperty={selectedProperty}
          selectedCountry={selectedCountry}
          // @ts-ignore
          propertyList={this.getPropertyList()}
          countryList={this.getCountryList()}
          onPropertyChange={this.onPropertyChange}
          onCountryChange={this.onCountryChange}
        />
        <FinanceOverview />
        <DuesContainer />
        <RemindersContainer />
        <TransactionsContainer toggleLoader={this.toggleLoading} isFromPortfolio={false} />
        {this.renderAddSheet()}
      </UserScreen>
    );
  };

  private renderAddSheet = (): React.ReactElement => {
    const { showAddSheet } = this.state;
    const data = this.getAddSheetData();

    return (
      <IconSheet
        data={data}
        sheetHeight={350}
        isVisible={showAddSheet}
        onCloseSheet={(): void => this.toggleAddSheet(false)}
      />
    );
  };

  private onPropertyChange = (propertyId: number): void => {
    const { selectedProperty, setCurrentProperty } = this.props;
    if (selectedProperty === propertyId) {
      return;
    }
    setCurrentProperty(propertyId);
  };

  private onCountryChange = (countryId: number): void => {
    const { selectedCountry, setCurrentCountry } = this.props;
    if (selectedCountry === countryId) {
      return;
    }
    setCurrentCountry(countryId);
  };

  private onAddRecord = (): void => {
    const {
      assets,
      t,
      navigation: { navigate },
    } = this.props;

    if (assets.length <= 0) {
      AlertHelper.error({ message: t('addProperty') });
      return;
    }

    navigate(ScreensKeys.AddRecordScreen);
  };

  private onSetReminder = (): void => {
    const {
      navigation: { navigate },
    } = this.props;

    navigate(ScreensKeys.AddReminderScreen);
  };

  private getAddSheetData = (): ISheetData[] => {
    const {
      t,
      navigation: { navigate },
    } = this.props;
    const iconSize = 40;
    const ImageHOC = (Image: React.FC<SvgProps>): React.ReactElement => <Image width={iconSize} height={iconSize} />;
    const IconHOC = (name: string): React.ReactElement => (
      <Icon name={name} size={iconSize + 3} color={theme.colors.error} />
    );
    return [
      {
        icon: ImageHOC(Accounting),
        label: t('assetFinancial:addFinancials'),
        onPress: this.onAddRecord,
      },
      {
        icon: IconHOC(icons.reminder),
        label: t('assetFinancial:setReminders'),
        onPress: this.onSetReminder,
      },
      {
        icon: ImageHOC(Bank),
        label: t('assetFinancial:bankAccounts'),
        onPress: (): void => navigate(ScreensKeys.AddBankAccount),
      },
      {
        icon: ImageHOC(Payment),
        label: t('propertyPayment:propertyPayment'),
        onPress: (): void => navigate(ScreensKeys.PropertyPayment),
      },
    ];
  };

  private toggleAddSheet = (isVisible: boolean): void => {
    this.setState({ showAddSheet: isVisible });
  };

  private toggleLoading = (loading: boolean): void => {
    this.setState({ isLoading: loading });
  };

  private getHeaderData = (): IMetricsData[] => {
    const {
      t,
      ledgerMetrics: { income, expense },
    } = this.props;
    const currentYear = DateUtils.getCurrentYear();

    return [
      {
        name: t('assetFinancial:income', { year: currentYear }),
        count: income,
        isCurrency: true,
        colorCode: theme.colors.incomeGreen,
      },
      {
        name: t('assetFinancial:expense', { year: currentYear }),
        count: expense,
        isCurrency: true,
        colorCode: theme.colors.yellowTint2,
      },
    ];
  };

  private getCountryList = (): Country[] => {
    const { assets } = this.props;
    return uniqBy(
      assets.map((asset) => asset.country),
      'id'
    );
  };

  private getPropertyList = (): PickerItemProps[] => {
    const { assets, selectedCountry } = this.props;

    // @ts-ignore
    return (selectedCountry === 0 ? assets : assets.filter((asset) => selectedCountry === asset.country.id)).map(
      (asset) => ({
        label: asset.projectName,
        value: asset.id,
      })
    );
  };
}

const mapStateToProps = (state: IState): IStateProps => {
  const { getUserAssets } = UserSelector;
  const { getFinancialLoaders, getLedgerData, getSelectedCountry, getSelectedProperty, getLedgerMetrics } =
    FinancialSelectors;
  return {
    assets: getUserAssets(state),
    financialLoaders: getFinancialLoaders(state),
    ledgerData: getLedgerData(state),
    selectedProperty: getSelectedProperty(state),
    selectedCountry: getSelectedCountry(state),
    ledgerMetrics: getLedgerMetrics(state),
  };
};

const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const {
    getLedgers,
    setCurrentCountry,
    setCurrentProperty,
    setTimeRange,
    getLedgerMetrics,
    resetLedgerFilters,
    setCurrentDueId,
    setCurrentReminderId,
    clearReminderFormData,
  } = FinancialActions;
  const { getAssetUsersSuccess } = AssetActions;
  const { resetPaymentState } = PropertyPaymentActions;
  return bindActionCreators(
    {
      getLedgers,
      setCurrentCountry,
      setCurrentProperty,
      setTimeRange,
      getLedgerMetrics,
      resetLedgerFilters,
      setCurrentDueId,
      setCurrentReminderId,
      getAssetUsersSuccess,
      clearReminderFormData,
      resetPaymentState,
    },
    dispatch
  );
};

const connectedComponent = connect(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation(LocaleConstants.namespacesKey.assetFinancial)(Financials));
export default connectedComponent;
