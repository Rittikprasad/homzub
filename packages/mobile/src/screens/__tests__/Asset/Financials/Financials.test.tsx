import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Financials } from '@homzhub/mobile/src/screens/Asset/Financials';

describe.skip('Financials Screen', () => {
  let component: ShallowWrapper;
  let props: any;

  beforeEach(() => {
    props = {
      navigation: {
        goBack: jest.fn(),
        addListener: jest.fn(),
      },
    };
    component = shallow(<Financials {...props} t={(key: string): string => key} />);
  });

  it('should render snapshot', () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
