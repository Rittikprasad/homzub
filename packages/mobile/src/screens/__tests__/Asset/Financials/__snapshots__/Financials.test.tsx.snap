// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Financials Screen should render snapshot 1`] = `undefined`;
